codegenerator/cli/templates/static/codegen/src/eventFetching/FetchState.res

```diff
@@ -185,262 +84,509 @@ Pass the shorter list into A for better performance
 */
 let mergeSortedEventList = (a, b) => Utils.Array.mergeSorted(eventCmp, a, b)
 
-let mergeWithNextRegister = (register: register, ~next: register) => {
-  let fetchedEventQueue = mergeSortedEventList(register.fetchedEventQueue, next.fetchedEventQueue)
-  let contractAddressMapping = ContractAddressingMap.combine(
-    register.contractAddressMapping,
-    next.contractAddressMapping,
-  )
+let mergeIntoPartition = (register: register, ~target: register, ~maxAddrInPartition) => {
+  let latestFetchedBlock = target.latestFetchedBlock
 
-  let dynamicContracts = DynamicContractsMap.merge(register.dynamicContracts, next.dynamicContracts)
+  let mergedContractAddressMapping = target.contractAddressMapping->ContractAddressingMap.copy
+  let mergedDynamicContracts = target.dynamicContracts->Js.Array2.copy
 
-  if register.latestFetchedBlock.blockNumber !== next.latestFetchedBlock.blockNumber {
-    Js.Exn.raiseError("Invalid state: Merged registers should belong to the same block")
-  }
+  let restDcsCount =
+    target.contractAddressMapping->ContractAddressingMap.addressCount +
+    register.contractAddressMapping->ContractAddressingMap.addressCount -
+    maxAddrInPartition
 
-  {
-    id: next.id,
-    fetchedEventQueue,
-    contractAddressMapping,
-    dynamicContracts,
-    firstEventBlockNumber: Utils.Math.minOptInt(
-      register.firstEventBlockNumber,
-      next.firstEventBlockNumber,
-    ),
-    latestFetchedBlock: next.latestFetchedBlock,
+  let rest = if restDcsCount > 0 {
+    let restAddresses = Utils.Set.make()
+
+    let restDcs = register.dynamicContracts->Js.Array2.slice(~start=0, ~end_=restDcsCount)
+    restDcs->Array.forEach(dc => {
+      let _ = restAddresses->Utils.Set.add(dc.contractAddress)
+    })
+
+    let restContractAddressMapping = ContractAddressingMap.make()
+
+    register.contractAddressMapping.nameByAddress
+    ->Js.Dict.keys
+    ->Belt.Array.forEach(key => {
+      let name = register.contractAddressMapping.nameByAddress->Js.Dict.unsafeGet(key)
+      let address = key->Address.unsafeFromString
+      let map =
+        restAddresses->Utils.Set.has(address)
+          ? restContractAddressMapping
+          : mergedContractAddressMapping
+      map->ContractAddressingMap.addAddress(~address, ~name)
+    })
+
+    let _ =
+      mergedDynamicContracts->Js.Array2.pushMany(
+        register.dynamicContracts->Js.Array2.sliceFrom(restDcsCount),
+      )
+
+    Some({
+      id: register.id,
+      status: {
+        isFetching: false,
+      },
+      fetchedEventQueue: [],
+      contractAddressMapping: restContractAddressMapping,
+      dynamicContracts: restDcs,
+      latestFetchedBlock,
```

Comment:
Should there be any validation in this function that latestFetchedBlock is exactly the same between the partitions being merged?

Reply:
I thought it's not necessary, since the function won't run when the latestFetchedBlock is not the same (according to the setQueryResponse logic). And this is checked by integration tests.